src/sentry/migrations/0866_backfill_open_periods.py

```diff
@@ -0,0 +1,217 @@
+            end_activity = activity
+        elif activity.type == ActivityType.SET_REGRESSION.value:
+            start = activity.datetime
+            if end is None:
+                logger.error(
+                    "No end activity found for group open period backfill",
+                    extra={"group_id": group_id, "starting_activity": activity.id},
+                )
+            if start is not None and end is not None:
+                open_periods.append(
+                    GroupOpenPeriod(
+                        group_id=group_id,
+                        project_id=project_id,
+                        date_started=start,
+                        date_ended=end,
+                        resolution_activity=end_activity,
+                        user_id=end_activity.user_id if end_activity else None,
+                    )
+                )
+                end = None
+                end_activity = None
+    # Add the very first open period, which has no UNRESOLVED activity for the group creation
+    open_periods.append(
+        GroupOpenPeriod(
+            group_id=group_id,
+            project_id=project_id,
+            date_started=first_seen,
+            date_ended=end if end else None,
+            resolution_activity=end_activity,
+            user_id=end_activity.user_id if end_activity else None,
+        )
+    )
+    return open_periods
+
+
+def _backfill_group_open_periods(
+    apps: StateApps, group_data: list[tuple[int, datetime, int, int]]
+) -> None:
+    GroupOpenPeriod = apps.get_model("sentry", "GroupOpenPeriod")
+    Activity = apps.get_model("sentry", "Activity")
+
+    old_open_periods = list(
+        GroupOpenPeriod.objects.filter(group_id__in=[group_id for group_id, _, _, _ in group_data])
+        .values("group_id")
+        .annotate(Max("date_started"))
+    )
+    activities = list(
+        Activity.objects.filter(
+            group_id__in=[group_id for group_id, _, _, _ in group_data],
+            type__in=[ActivityType.SET_REGRESSION.value, ActivityType.SET_RESOLVED.value],
+        ).order_by("-datetime")
+    )
+
+    open_periods = []
+    for group_id, first_seen, status, project_id in group_data:
+        # Skip groups that already have open periods starting from the first seen date.
+        # These groups were either already backfilled or were created after the open period
+        # logic was added.
+        old_open_period = next(
+            (open_period for open_period in old_open_periods if open_period.group_id == group_id),
+            None,
+        )
+        if old_open_period and old_open_period.date_started == first_seen:
+            continue
+        query_end = old_open_period.date_started if old_open_period else None
+
+        activities_for_group = [
+            activity
+            for activity in activities
+            if activity.group_id == group_id
+            and (query_end is None or activity.datetime < query_end)
+        ]
```

Comment:
It might be better to do this once outside of this loop, and just have it be a lookup dict of `group_id -> list[valid_activity_rows]`